pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -1334,6 +1334,11 @@ func updatePipelineRunStatusFromChildObjects(ctx context.Context, logger *zap.Su
 	fullEmbedded := cfg.FeatureFlags.EmbeddedStatus == config.FullEmbeddedStatus || cfg.FeatureFlags.EmbeddedStatus == config.BothEmbeddedStatus
 	minimalEmbedded := cfg.FeatureFlags.EmbeddedStatus == config.MinimalEmbeddedStatus || cfg.FeatureFlags.EmbeddedStatus == config.BothEmbeddedStatus
 
+	if cfg.FeatureFlags.EmbeddedStatus == config.MinimalEmbeddedStatus {
+		pr.Status.TaskRuns = nil
+		pr.Status.Runs = nil
+	}
```

Comment:
I'm not sure I understand this change. Why are `pr.status.taskruns` and `pr.status.runs` set in the first place if embedded-status is "minimal"?

Is the change just from an empty slice to a nil slice? This is my interpretation based on the tests. If that's the case, we should change the tests to not care whether an empty slice or a nil slice is used.

If these fields are set incorrectly (i.e. to something other than an empty slice or nil pointer), can we change the part of the code that sets them, add tests, and link a bug that gives more context on the problem being encountered?

Reply:
Yeah, it's empty map to nil map. I remember having weird hassles with those maps being empty sometimes and nil sometimes when I added the minimal status in the first place, but can't remember details.

Reply:
> Why are pr.status.taskruns and pr.status.runs set in the first place if embedded-status is "minimal"?
In the 'normal' cases, it should be the case as described. But since the context was switching the defaultEmbeddedStatus from `full` to `minimal`, there could be edge cases where an existing piplineRun has `.stats.taskruns` or `.status.runs`. So in this case, it would error out at https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun.go#L1362 or https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun.go#L1362

Reply:
This case happened to cover this case:
https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun_test.go#L1891. But I am not sure if a similar one shall be added to this PR as the 'DefaultEmbeddedStatus`  to minimal would be in #5934 . If preferred, I can add this edge case in that PR?

Reply:
> Why are pr.status.taskruns and pr.status.runs set in the first place if embedded-status is "minimal"?

Great question, this could be a very rare edge case that @JeromeJu  wants to catch: say a PipelienRun contains a TaskRun that could run for a period time. While the TaskRun is running, we switched the embedded status feature flag from `full` to `minimum`, at this time, `pr.status.runs` are set while embedded-status is "minimal".

Reply:
Thanks @lbernick , updated with the comment. But I feel like setting them as `nil` rather than having any empty `status.runs` or `status.taskruns` would be more appropriate since we are going to remove them eventually according to TEP100?

Reply:
Can you elaborate a bit? having `status.runs` set to nil is meaningfully the same as having it set to an empty map. No client should be differentiating between the two cases, which is why I suggest our tests do the same. I'm not sure why setting these fields to nil would make the transition to removing these fields easier.

Also, do you think it is possible to write an integration test for this scenario?

Reply:
> Hm I'm wondering if it would be possible/worthwhile to try to write an integration test for this scenario?

I would like to confirm, by an integration test, what are the scenarios that need to be covered?

I think for the case where there could be a pipelineRunStatus with existing `status.runs` or `status.taskruns` could have been covered by [TestReconcileForCustomTaskWithPipelineRunTimedOut](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun_test.go#L1891). I am willing to create a thorough one if there're more complicated scenarios 🤔

Reply:
It sounds like the goal of this PR is to avoid errors in the case where someone changes the value of embedded-status while a pipelineRun is running. How do we know if it works or not?

Reply:
> It sounds like the goal of this PR is to avoid errors in the case where someone changes the value of embedded-status while a pipelineRun is running. How do we know if it works or not?

Agreed. Then I think the integration test shall be added in the same PR as https://github.com/tektoncd/pipeline/pull/5934?

And I am trying to picture the cases to be covered: probably one PipelineRun with all the fields in `status.taskruns` being switched from `Full` or `Both` to `Minimal` shall be sufficient?

Reply:
> Can you elaborate a bit? having status.runs set to nil is meaningfully the same as having it set to an empty map. No client should be differentiating between the two cases, which is why I suggest our tests do the same. I'm not sure why setting these fields to nil would make the transition to removing these fields easier.

I hope the following makes sense to elaborate:

- After the `PipelineRunStatus` migration:
  The existence of `status.runs{}` and `status.taskruns{}` could suggest using them in some ways. I think removing them entirely could be mitigating the chances for those to be exposed to users.
- During the switch of the `DefaultEmbeddedStatus`:
  There should not be `status.taskruns` or `status.runs` populated if embedded-status is "minimal", removing them could also be a guarantee that these fields are not existent at all when defaulting to minimal.
- For now:
  It could avoid potential hassles of `status.taskruns` or `status.runs` as suggested

Reply:
> This should be in our code standards, but tests should be included in the same pull request as the code that's being tested.

Thanks Lee! Got it.

Reply:
> * After the `PipelineRunStatus` migration:
>   The existence of `status.runs{}` and `status.taskruns{}` could suggest using them in some ways. I think removing them entirely could be mitigating the chances for those to be exposed to users.
> * During the switch of the `DefaultEmbeddedStatus` to `minimal`:
>   There should not be `status.taskruns` or `status.runs` populated if embedded-status is "minimal", removing them could also be a guarantee that these fields are not existent at all when defaulting to minimal.
> * For now:
>   It could avoid potential hassles of `status.taskruns` or `status.runs` as suggested

I'm still a bit confused-- I'm not too concerned that users will think they should be using taskruns and runs if set to an empty map, when childrefs are populated. I think it's ok for taskruns and runs to be set to an empty map if embedded status is minimal.

It still sounds like the root problem is concern about users migrating to minimal embedded status experiencing bugs with their pipelinerun. Can you try to reproduce this issue on your own cluster, and if you're able to reproduce it, file a bug?

Based on the previous conversation on this PR, it seems like this PR might have been intended to fix a test failure on https://github.com/tektoncd/pipeline/pull/5934? If that's the case, you can fix this by using `cmp.Diff(want, got, cmpopts.EquateEmpty()`.

Reply:
Thanks Lee for the reminder, created https://github.com/tektoncd/pipeline/issues/5991. Could you help take a look if that makes sense for addressing the issue that's intended for this PR?

> Based on the previous conversation on this PR, it seems like this PR might have been intended to fix a test failure on https://github.com/tektoncd/pipeline/pull/5934? If that's the case, you can fix this by using cmp.Diff(want, got, cmpopts.EquateEmpty().

I don't think that case could be fixed with `cmp.Diff(want, got, cmpopts.EquateEmpty().`

Reply:
#5991 won't be fixed with `cmpopts.EquateEmpty` but I think it's still best practice and worth adding here.